src/transport/PointToPointBroker.cpp

```diff
@@ -566,10 +672,74 @@ std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
 
     Message message = recvEndpoints[label]->recv();
 
-    // TODO - possible to avoid this copy?
     return message.dataCopy();
 }
 
+std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
+                                                     int sendIdx,
+                                                     int recvIdx)
+{
+    // If we don't need to receive messages in order, return here
+    if (!orderMsg.load(std::memory_order_acquire)) {
+        return doRecvMessage(groupId, sendIdx, recvIdx);
+    }
+
+    int expectedSeqNum = getExpectedSeqNum(groupId, sendIdx);
+    std::vector<uint8_t> recvData;
+    std::vector<uint8_t> returnData;
+
+    // Loop to receive the right sequence number
+    while (true) {
+        // First, receive from the transport layer
+        recvData = doRecvMessage(groupId, sendIdx, recvIdx);
+
+        // Second, get the sequence number from the received message
+        int seqNum;
+        faabric::util::readBytesOf(
+          recvData, recvData.size() - sizeof(int), &seqNum);
+        recvData.resize(recvData.size() - sizeof(int));
+
+        // If the sequence numbers match, exit the loop
+        if (seqNum == expectedSeqNum) {
+            returnData = std::move(recvData);
+            break;
+        }
+
+        // If not, we must insert the received message in the out of order
+        // received messages. We must also check if the sequence number we want
+        // is in the list.
+
+        // First insert
+        auto insertIt = std::find_if(
+          outOfOrderMsgs.begin(),
+          outOfOrderMsgs.end(),
+          [&seqNum](const std::pair<int, std::shared_ptr<std::vector<uint8_t>>>&
+                      element) { return element.first > seqNum; });
+        outOfOrderMsgs.emplace(
+          insertIt,
+          std::make_pair(seqNum,
+                         std::make_shared<std::vector<uint8_t>>(recvData)));
```

Comment:
Double check, but I think this is a copy, e.g. if you change `recvData`, the content of the pointer target won't change.